exchanges/stream/websocket.go

```diff
@@ -283,27 +321,139 @@ func (w *Websocket) Connect() error {
+			multiConnectFatalError = fmt.Errorf("cannot connect to [conn:%d] [URL:%s]: %w ", i+1, w.connectionManager[i].Setup.URL, errWebsocketSubscriptionsGeneratorUnset)
+			break
+		}
+
+		subs, err := w.connectionManager[i].Setup.GenerateSubscriptions() // regenerate state on new connection
 		if err != nil {
-			log.Errorf(log.WebsocketMgr, "%s cannot start websocket connection monitor %v", w.GetName(), err)
+			if errors.Is(err, asset.ErrNotEnabled) {
+				if w.verbose {
+					log.Warnf(log.WebsocketMgr, "%s websocket: %v", w.exchangeName, err)
+				}
+				continue // Non-fatal error, we can continue to the next connection
+			}
+			multiConnectFatalError = fmt.Errorf("%s websocket: %w", w.exchangeName, common.AppendError(ErrSubscriptionFailure, err))
+			break
+		}
+
+		if len(subs) == 0 {
+			// If no subscriptions are generated, we skip the connection
+			if w.verbose {
+				log.Warnf(log.WebsocketMgr, "%s websocket: no subscriptions generated", w.exchangeName)
+			}
+			continue
+		}
+
+		if w.connectionManager[i].Setup.Connector == nil {
+			multiConnectFatalError = fmt.Errorf("cannot connect to [conn:%d] [URL:%s]: %w ", i+1, w.connectionManager[i].Setup.URL, errNoConnectFunc)
+			break
+		}
+		if w.connectionManager[i].Setup.Handler == nil {
+			multiConnectFatalError = fmt.Errorf("cannot connect to [conn:%d] [URL:%s]: %w ", i+1, w.connectionManager[i].Setup.URL, errWebsocketDataHandlerUnset)
+			break
+		}
+		if w.connectionManager[i].Setup.Subscriber == nil {
+			multiConnectFatalError = fmt.Errorf("cannot connect to [conn:%d] [URL:%s]: %w ", i+1, w.connectionManager[i].Setup.URL, errWebsocketSubscriberUnset)
+			break
+		}
+
+		// TODO: Add window for max subscriptions per connection, to spawn new connections if needed.
+
+		conn := w.getConnectionFromSetup(w.connectionManager[i].Setup)
+
+		err = w.connectionManager[i].Setup.Connector(context.TODO(), conn)
+		if err != nil {
+			multiConnectFatalError = fmt.Errorf("%v Error connecting %w", w.exchangeName, err)
+			break
 		}
+
+		w.Wg.Add(1)
+		go w.Reader(context.TODO(), conn, w.connectionManager[i].Setup.Handler)
+
+		w.connections[conn] = &w.connectionManager[i]
+
+		err = w.connectionManager[i].Setup.Subscriber(context.TODO(), conn, subs)
+		if err != nil {
+			multiConnectFatalError = fmt.Errorf("%v Error subscribing %w", w.exchangeName, err)
+			break
+		}
+
+		if w.verbose {
+			log.Debugf(log.WebsocketMgr, "%s websocket: [conn:%d] [URL:%s] connected. [Subscribed: %d]",
+				w.exchangeName,
+				i+1,
+				conn.URL,
+				len(subs))
+		}
+
+		w.connectionManager[i].Connection = conn
 	}
 
-	subs, err := w.GenerateSubs() // regenerate state on new connection
-	if err != nil {
-		return fmt.Errorf("%s websocket: %w", w.exchangeName, common.AppendError(ErrSubscriptionFailure, err))
+	if multiConnectFatalError != nil {
+		// Roll back any successful connections and flush subscriptions
+		for conn, candidate := range w.connections {
```

Comment:
Unless there is a reason otherwise, we shouldn't ever really iterate over the map if the slice contains the same fields. Should you not be using `w.connectionManager[x].Connection.Shutdown`?